hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java

```diff
@@ -382,6 +395,14 @@ public void deleteSnapshot(final INodesInPath iip, final String snapshotName,
             EnumSet.of(XAttrSetFlag.CREATE, XAttrSetFlag.REPLACE));
         return;
       }
+
```

Comment:
I feel all the deletion should happen via one code path, i.e. the background thread. some deletion from the user thread and others from background is error prone.

Reply:
@jnp , what is your thought?

Reply:
IMO, if the user is delete the first snapshot.  We should do it right away.  Otherwise, there is no way for the user to free up the resource from the snapshots.

Reply:
I agree with @szetszwo here. If the user follows the order of deletion, it doesn't make sense to delay it further by pushing it to a background thread. Snapshot deletion is still happening by the same code path in SnapshotManager but the triggers are different.

SnapshotDeletiongc as well as User deleteSnapshotOperation directly calls FsDirSnapshot#deleteSnapshot() which eventually calls SnapshotManger function while Replay of edits will directly call SnapshotManager#deleteSnapshot bypassing all the checks. So, essentially every deletion in the system won't be taking the exact path.

Reply:
I agree that code paths are similar. However, just wanted to discuss this once. :) The only thing which might not work is the sleep between every snap deletes here. It may happen that the user operation snap delete and the background thread based snap delete are not time spaced correctly. But that is not a big issue anyways.

Reply:
Sure, thanks.